chore: add MIRACL&MrTydi, GPU support, Colbert evals

bclavie opened the pull request:
おはよう！

I'm opening this PR to upstream some of the changes which have been really helpful for me. It's a big PR (my apologies!), but the changes are broadly in three categories:

**Datasets**
- Add support (easy to comment out) for MIRACL and Mr.TyDi, Japanese subplot.
- To avoid running it on the entire 6-7M doc corpuses for each, negatives are sampled
  - Randomly for MrTyDi (parameter to decide how many negs to fetch)
  - Using the list of hard negatives from @oshizo's [embedding evaluation repository](https://github.com/oshizo/JapaneseEmbeddingEval) for MIRACL, or randomly (parameter can be used to decide, defaults to hard negatives). The `corpus_sample` parameter decides how many hard negative per examples to use if using `hard_negative`, or how many total random examples to fetch if using `random`.

**GPU Embeddings**
- If cuda is present on the machine, automatically default to using GPU to speed up evaluation.

**Support to evaluate late-interaction retrievers**
- Late Interactions models like SPLADE++ and ColBERT are increasingly powerful approaches to embedding for document retrieval. This PR tries to mimic the existing style of the repo to add support for ColBERT-style retrieval.
-  This also required a slight update to the recall calculation function, to make sure it can accept either scores or already-generated lists of y_pred.


I hope you find this useful

bclavie commented:
@TatsuyaShirakawa I've just pushed an update to include the results of the initial tentative release of [JaColBERT](https://huggingface.co/bclavie/JaColBERT) as well as the results of every other non-API modesl on all three of the benchmarks, I hope this is okay with you! Please let me know if there's any issue.

TatsuyaShirakawa commented on configs/default.yml:
```diff
@@ -10,10 +19,10 @@ embeddings:
-    model_name: textembedding-gecko-multilingual@001
-  - name: VertexAITextEmbedding
-    model_name: textembedding-gecko-multilingual@latest
+  # - name: VertexAITextEmbedding
```
@bclavie 
I wanna keep this not commented-out (the same goes for other parts).
If you want to run with this not working, how about create your own config by copying this default.yml?

bclavie replied:
Good catch! I have uncommented all the models, I forgot to change default.yml back after running my tests.

TatsuyaShirakawa commented on dvsb/late_interaction_retrievers/colbert_.py:
Why the filename is `colbert_.py` instead of `_colbert.py`?
If the `_` implies private something, I would like to follow [Google style guides](https://google.github.io/styleguide/pyguide.html#3163-file-naming).

bclavie replied:
The `_` is just to avoid clashing with the default `colbert` from the `colbert-ir` library. That does essentially make it a private version of it, so you're absolutely right that it should be a leading `_` instead. Change pushed!

TatsuyaShirakawa commented:
@bclavie Thank you for reaching out to this repo!
And thank you for adding your great work! 🎉 

I left some comments on the code. Please take a brief look 🙏

TatsuyaShirakawa commented on dvsb/metric/metric.py:
```diff
@@ -11,7 +11,12 @@ def get_name(self) -> str:
+    def compute(
+        self,
+        y_true: list[list[int]],
+        scores: npt.NDArray[np.float64] | None = None,
```
Using `|` for creating union is only supported Python >= 3.10 but I would like to support Python 3.9 as well.
Can you replace this with `Optional[npt.NDArray[np.float64]]`? (same goes for other parts)

bclavie replied:
Oh, my bad, I'm too used to the >= 3.10 style of typing now! It's changed to `Optional[...]`.

bclavie commented:
I've made the changes you suggested, thanks for taking a look and spotting the issues!

TatsuyaShirakawa commented on configs/default.yml:
```diff
@@ -2,14 +2,23 @@ datasets:
+   model_name: embed-multilingual-light-v3.0
+ - name: CohereEmbedding
+   model_name: embed-multilingual-v3.0
   - name: VertexAITextEmbedding
```
I tried VertexAITextEmbedding but, with MrTyDi dataset, token length seems too long for VertexAITextEmbedding.
![image](https://github.com/nouu-me/document_vector_search_benchmark/assets/11359334/8c37bf34-92cd-46dc-aa93-3a3fd8f10a94)

How about commenting out all of "cloud embeddings" (Cohere, VertexAI, OpenAI)?

TatsuyaShirakawa commented:
@bclavie Sorry for late response! I tried running the code but found some errors (above). Please take a look when you're free 🙏

bclavie commented:
No problem, thank you very much for responding! I will fix the spacing issue immediately! 

As for the length limit for the cloud embeddings, would you rather I:
- Modify the code so it's embedded in batches small enough for the remote providers
- Comment them out as you suggested?

Finally, there is an issue with using JAColBERT via the original `colbert-ir` lib if installed in ways other than directly getting it from source, which I'm hoping to fix upstream in the ColBERT repo itself (https://github.com/stanford-futuredata/ColBERT/pull/282), but it might be worth waiting until then to merge this PR! My apologies for this.

TatsuyaShirakawa commented:
@bclavie Thanks for you suggestion

> As for the length limit for the cloud embeddings, would you rather I:
> 
> - Modify the code so it's embedded in batches small enough for the remote providers
> - Comment them out as you suggested?

I think the former one (small batch) is better for this case because this could not be a issue limited to cloud APIs.

bclavie commented:
Hey @TatsuyaShirakawa, my apologies, I forgot to post here. I did have a working fix for the context length! I'll send it over in a bit.

Either today or tomorrow I will also:
- Modify the benchmark so TyDi and MIRACL do not use random documents but the top 1000 BM25 results for each query.  This is better practice (even some papers benchmark this way!) 
- Run the benchmark for JaColBERTv2.
- Hopefully re-run the benchmark for every other embedding so we get metrics on a better quality eval set than the current TyDi/MIRACL implementations.

TatsuyaShirakawa commented:
Thanks @bclavie for your great contribution!
I'm keenly interested in the performacne of JaColBERTv2!

bclavie commented:
It's all functional now! I'm running JSQuAd + MIRACL (w/ top 1000 BM25 passages for each query) locally for all (non-API) models and will update the PR once that's done! TyDi and MIRACL are pretty similar so I'm just doing one of them to make the table more readable and save some compute.

JaColBERTv2 seems state-of-the-art on generalisation benchmarks (JSQuAd, JQaRA) and competitive even out-of-domain! In the results I have so far it does better on MIRACL than all models except BGE-M3 (trained on it) 😄

bclavie commented:
@TatsuyaShirakawa 🎉 After much delay, the PR is finally done!

Main changes:
- Includes JaColBERTv1, JaColBERTv2, and BGE-M3 (dense vector only)
- Evals on JSQuAd (R@{1,3,5,10}) and MIRACL (R@{3,5,10, 100}), using a dataset composed of the top 1000 BM25 results for each query.
- Did not run full evals on TyDi as it's pretty similar to MIRACL and was making the table nearly unreadable, but it's implemented and the top 1000 BM25 candidates are available.
- Updated the table with the results from the full run
- Changed batch sizes so they're adapted to each model, including the remote ones, and added a dynamic sequence length detector so they all get set to 512 unless manually specified in the config.
- Set device to `cuda` if cuda is detected, otherwise defaults to CPU.
- Modified the CSV to Markdown table to group by Dataset, and drop the columns that are less informative (R@1 is useless for MIRACL and so is R@100 for JSQuAd)

TatsuyaShirakawa commented:
@bclavie I finally merged your PR 😃 
Again, thank you for your excellent work!
